### PR TITLE
Add meta.mainProgram to writeCBin, add tests

### PR DESCRIPTION
**Pull request by roberth**
###### Description of changes

- Make explicit that the generated packages have a main program.
- Add tests
- Put the test list in a better location

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
@ofborg build tests.trivial-builders

**Review by ncfavier — APPROVED** [review]

File: pkgs/build-support/trivial-builders/default.nix [hunk 1]
```diff
@@ -242,7 +242,14 @@ rec {
 
 
   */
-  writeScriptBin = name: text: writeTextFile {inherit name text; executable = true; destination = "/bin/${name}";};
+  writeScriptBin = name: text: writeTextFile {
```
Comment (figsoda) [inline pkgs/build-support/trivial-builders/default.nix:245]: #234454 should be able to handle this

File: pkgs/build-support/trivial-builders/default.nix [hunk 2]
```diff
@@ -300,6 +307,9 @@ rec {
       checkPhase = ''
         ${stdenv.shellDryRun} "$target"
       '';
+      meta = {
```
Comment (figsoda) [inline pkgs/build-support/trivial-builders/default.nix:310]: same for this

**Comment by figsoda** [comment]
Would be nice if edge cases like https://github.com/NixOS/nixpkgs/pull/234072#issuecomment-1563557992 were also tested, since `lib.getExe` already falls back to parsing `name`

File: pkgs/build-support/trivial-builders/default.nix [hunk 3]
```diff
@@ -242,7 +242,14 @@ rec {
...
 
   */
-  writeScriptBin = name: text: writeTextFile {inherit name text; executable = true; destination = "/bin/${name}";};
+  writeScriptBin = name: text: writeTextFile {
+    inherit name text;
+    executable = true;
+    destination = "/bin/${name}";
+    meta = {
+      mainProgram = name;
+    };
```
Comment (AndersonTorres) [inline pkgs/build-support/trivial-builders/default.nix:251]: If there is no other meta attrs,


```suggestion
    meta.mainProgram = name;
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by roberth** [comment]
> Would be nice if edge cases like [#234072 (comment)](https://github.com/NixOS/nixpkgs/pull/234072#issuecomment-1563557992) were also tested, since `lib.getExe` already falls back to parsing `name`

@figsoda, does [this change](https://github.com/NixOS/nixpkgs/compare/2201cd9f3092fcf930633090f144d7c0e46aa119..a6aabd9d28ea935072b4825bbc15ef7cc40e5e36) cover what you had in mind?

**Comment by roberth** [comment]
> > Would be nice if edge cases like [#234072 (comment)](https://github.com/NixOS/nixpkgs/pull/234072#issuecomment-1563557992) were also tested, since `lib.getExe` already falls back to parsing `name`
> 
> @figsoda, does [this change](https://github.com/NixOS/nixpkgs/compare/2201cd9f3092fcf930633090f144d7c0e46aa119..a6aabd9d28ea935072b4825bbc15ef7cc40e5e36) cover what you had in mind?

Rereading, I didn't think it did.
I've added the test case for you, but that should really have been done in the original PR. Will appreciate that next time.

File: pkgs/build-support/trivial-builders/default.nix (hunk 3 above)
Comment (AndersonTorres) [reply pkgs/build-support/trivial-builders/default.nix]: I believe there is no reason to extend this in the future.

Will anyone take `maintainer`ship of these?

They don't run in all `platforms`?

They don't even need `homepage` or `changelog`.

On the other hand, a short `description` is certainly useful, but you didn't bother...

**Review by figsoda — APPROVED** [review]
tests lgtm at a glance

**Comment by figsoda** [comment]
> I've added the test case for you, but that should really have been done in the original PR. Will appreciate that next time.

:+1:

**Pull request merged**